packages/patternfly-4/react-table/src/components/Table/Table.tsx

```diff
@@ -0,0 +1,298 @@
+  props?: any;
+}
+
+export interface IRow extends RowType {
+  cells?: (React.ReactNode | IRowCell)[];
+  isOpen?: boolean;
+  parent?: number;
+  compoundParent?: number;
+  props?: any;
+  fullWidth?: boolean;
+  noPadding?: boolean;
+  showSelect?: boolean;
+  isExpanded?: boolean;
+  isFirstVisible?: boolean;
+  isLastVisible?: boolean;
+  selected?: boolean;
+}
+
+export interface TableProps {
+  'aria-label'?: string;
+  children?: React.ReactNode;
+  className?: string;
+  variant?: 'compact';
+  borders?: boolean;
+  gridBreakPoint?: '' | 'grid' | 'grid-md' | 'grid-lg' | 'grid-xl' | 'grid-2xl';
+  sortBy?: ISortBy;
+  onCollapse?: OnCollapse;
+  onExpand?: OnExpand;
+  onSelect?: OnSelect;
+  onSort?: OnSort;
+  actions?: IActions;
+  actionResolver?: IActionsResolver;
+  areActionsDisabled?: IAreActionsDisabled;
+  header?: React.ReactNode;
+  caption?: React.ReactNode;
+  rowLabeledBy?: string;
+  expandId?: string;
+  contentId?: string;
+  dropdownPosition?: 'right' | 'left';
+  dropdownDirection?: 'up' | 'down';
+  rows: (IRow | string[])[];
+  cells: (ICell | string)[];
+  bodyWrapper?: Function;
+  rowWrapper?: Function;
+  role?: string;
+}
+
+export const TableContext = React.createContext({
+  headerData: null as any,
+  headerRows: null as IHeaderRow[],
+  rows: [] as (IRow | string[])[]
+});
+
+export class Table extends React.Component<TableProps, {}> {
+  static defaultProps = {
+    children: null as React.ReactNode,
+    className: '',
+    variant: null as TableVariant,
+    borders: true,
+    rowLabeledBy: 'simple-node',
+    expandId: 'expandable-toggle',
+    contentId: 'expanded-content',
+    dropdownPosition: DropdownPosition.right,
+    dropdownDirection: DropdownDirection.down,
+    header: undefined as React.ReactNode,
+    caption: undefined as React.ReactNode,
+    'aria-label': undefined as string,
+    gridBreakPoint: TableGridBreakpoint.gridMd,
+    role: 'grid'
+  }
+
+  isSelected = (row: IRow) => row.selected === true;
+
+  areAllRowsSelected = (rows: IRow[]) => {
+    if (rows === undefined || rows.length === 0) {
+      return false;
+    }
+    return rows.every(row => this.isSelected(row) || (row.hasOwnProperty('parent') && !row.showSelect));
+  };
+    
+  render(){
+    const {
+      'aria-label': ariaLabel,
+      caption,
+      header,
+      className,
+      gridBreakPoint,
+      onSort,
+      onSelect,
+      sortBy,
+      children,
+      actions,
+      actionResolver,
+      areActionsDisabled,
+      onCollapse,
+      onExpand,
+      rowLabeledBy,
+      dropdownPosition,
+      dropdownDirection,
+      contentId,
+      expandId,
+      variant,
+      rows,
+      cells,
+      bodyWrapper,
+      rowWrapper,
+      borders,
+      role,
+      ...props
+    } = this.props;
+
+    if (!ariaLabel && !caption && !header && role !== 'presentation') {
+      // tslint:disable-next-line:no-console
+      console.error('Table: Specify at least one of: header, caption, aria-label')
```

Comment:
This is interesting.. I wonder if there is a way to use TypeScript to enforce this.

We might be able to break the props down into separate interfaces where each one has one of the three props required and the other two optional (and another interface with all 3 of those props optional and `role` is specifically only allowed to be `'presentation'`. Then we could `export interface TableProps = TablePropsWithHeader | TablePropsWithCaption | TablePropsWithAriaLabel | TablePropsWithRolePresentation` (all of which extend `BaseTableProps` which doesn't have any of those 4 props). Maybe that's not worth it, but we could avoid having to do this check at runtime.

Inspiration: https://stackoverflow.com/a/40510700

Reply:
interesting... yea i mimicked react-core w/ this approach, but I'm not opposed to that type of change. I think static type checking and the runtime error are BOTH beneficial. Maybe a separate  followup issue or piggybacking on #2673 w/ this for now...